tideways/provision.sh

```diff
@@ -68,11 +71,19 @@ function restart_php() {
 
 function install_xhgui_frontend() {
     cp -f "${DIR}/nginx.conf" "/etc/nginx/custom-utilities/xhgui.conf"
+    declare -a packages=()
+    for version in "7.0" "7.1" "7.2" "7.3" "7.4"; do
+        if [[ $(command -v php$version) ]]; then
+            packages+=("php${version}-sqlite3")
+        fi
+    done
+    apt-get -y --allow-downgrades --allow-remove-essential --allow-change-held-packages -o Dpkg::Options::=--force-confdef -o Dpkg::Options::=--force-confnew install --fix-missing --fix-broken "${packages[@]}"
     if [[ ! -d "/srv/www/default/xhgui" ]]; then
         echo -e " * Git cloning xhgui from https://github.com/perftools/xhgui.git"
         cd /srv/www/default
         noroot git clone "https://github.com/perftools/xhgui.git" xhgui
```

Comment:
you should not install @master branch, if you want stability. use releases!

1. git clone -b 0.16.0 ...

or (better, as it can pick latest release):
2. `composer create-project --no-dev perftools/xhgui`

you may want to add `--keep-vcs --prefer-source`, if you want `.git` as well

Reply:
I was thinking indeed of a fixed release so we can test it better the various changes.

Reply:
I'd prefer the `composer` method as it simplifies updating in the future, if a 0.17 release were made and we were checking out 0.16 branch that means we then have to do a tonne of git command wrangling for something that should be a 1 liner composer command

Reply:
@glensc my understanding of `create-project` is that it would take `perftools/xhgui` and perform a hard fork locally, and we'd be stuck on that version of XHGui unless we took the nuclear option of deleting the folder on every single provision and recreating it with a stable tag. Either that or having to write a check to see what the latest version of `perftools/xhgui` is.

Is there an alternative to this that's simpler? I'm used to declaring a dependency and letting `composer` manage the dependency ( I see XHGUI as a dependency of VVV, not a template to create a brand new project from ). I'd prefer to have stock vanilla XHGUI unmodified save for the insertion of a config file as the idealised situation.

I'm also not seeing any form of `stable` branch, I've a strong distaste for manually writing out and maintaining a number in a shell script.

Reply:
@tomjn please get acquainted with the context first, the git clone or composer create is executed only in the code branch when the target path does not exist. with the --keep-vcs the outcome is same, a git checkout, with the difference that composer create project uses latest release from packagist (git tag with the biggest version), and plain git clone uses master branch from github.

as for the "else-branch", if target path exist, there's no similar solution. you could do some trickery finding what is the latest "git tag" and checkout that. perhaps github apis or via git fetch

the alternative is maintaining composer.json file, but be aware that if you also keep composer.lock then it contains a lock to specific version that you are against. you would need to composer require xhgui each time (to resolve to latest tag from packagist). composer upgrade xhgui might not work for you as xhgui is pre-1.0 semver release, so every 0.x version released there is "major" version that needs to be bumped manually.

also, from personal practice, I've found that maintaining composer.json with composer.lock and keeping up changes from composer.lock of upstream project is very difficult (merging two lock files)

what I tried to say with the initial post, is that by using "master" you can end up with a broken state, there's no guarantee that "master" even works. there's some guarantee that git tag is stable. or if it's broken it's fixed first with new patch release. I here offered only replace git clone with a composer create-project. it provides no change to the "else" branch where you have git pull and rebase against master

Reply:
> please get acquainted with the context first

There's no need to be rude. If the current mechanism for checking wether it should clone or pull is insufficient it can be replaced. I referred to the problem of updating to new stable versions of XHGUI. The goal is to always be on the latest stable version without having to manually change the provisioner every time a release happens, or nuking the folder on every provision and reinstalling. `create-project` does not give us that, neither do tagged versions

Reply:
you can retrieve latest tag using github apis and then use that. either composer require foo:$version, composer update foo:$version or git checkout $version doesn't really matter in this case.

here's the first result from a google search:
- https://gist.github.com/lukechilds/a83e1d7127b78fef38c2914c4ececc3c

there are tons of solutions, with different tools. you should get the idea how from there and implement your own using your tools (php json perhaps if `jq` not available)

aside, if you are using prestine repo, you should also use composer install, not composer update, to use project provided composer.lock

Reply:
> here's the first result from a google search:

While I appreciate that you're trying to help, there's no need to be passive aggressive or use snark.

I understand there are many alternative options, and was indicating your recommended installation methods do not allow for automated upgrades of stable versions. It was a false assumption on your part that we were unaware of those options and the means by which to do them.

I would suggest reading [our code of conduct](https://github.com/Varying-Vagrant-Vagrants/VVV/blob/develop/.github/CODE_OF_CONDUCT.md)

---

I suspect `composer` can be used to handle this with a `composer.json` such as this one:

```json
{
    "require": {
        "perftools/xhgui": "@stable",
        "composer/installers": "~1.0 || ~2.0"
    },
    "extra": {
        "installer-paths": {
            "/srv/www/default/xhgui": ["vendor/package"]
        }
    }
}
```

That particular example is untested so some adjustment may be necessary, possibly running `composer.json` in that directory to ensure that the `vendor` folder is in the correct location, I'm unsure at the moment

Reply:
Until then, I suggest we stick to 0.16.0 so we can unblock this PR